### ISSUE TITLE
`-fixfn` test hangs on OSX; reveals larger issue of float comparison

### ISSUE THREAD

**Issue by occidens** [magnars/dash.el#123]
Here is a interesting problem: I was running the test suite on my Mac, and it hung on the test of `-fixfn`. I then ran the tests in a Linux virtual machine with no problem.

Closer inspection reveals that there are some subtle platform-specific (or Emacs minor-version-specific?) differences in the floating-point math in the `cos` function. While the fixpoint of `cos` can be *verified* on both systems

```emacs-lisp
(= (cos 0.7390851332151607) 0.7390851332151607) ;; => t
```

On a Mac (tested on `GNU Emacs 24.4.1 x86_64-apple-darwin13.4.0, NS apple-appkit-1265.21`), `-fixfn` never converges on a solution because `cos` oscillates between *0.739&#x2026;8* and *0.739&#x2026;5*.

```emacs-lisp
(cos (cos 0.7390851332151608)) ;; => 0.7390851332151608

(cos (cos (cos 0.7390851332151608))) ;; => 0.7390851332151605

(cos (cos (cos (cos 0.7390851332151608)))) ;; => 0.7390851332151608
```

On Linux (tested on `GNU Emacs 24.4.91.1 (x86_64-unknown-linux-gnu, GTK+ Version 2.24.10)`), however, `cos` does converge to the value specified in the test suite.

```emacs-lisp
(cos (cos 0.7390851332151608)) ;; => 0.7390851332151607

(cos (cos (cos 0.7390851332151608))) ;; => 0.7390851332151607

(cos (cos (cos (cos 0.7390851332151608)))) ;; => 0.7390851332151607
```

As a temporary solution, I will submit a pull request in a moment for a change to `run-tests.sh` that will allow the test to be skipped.

To solve the larger issue, what do you think of the following solution, which would allow the caller of `-fixfn` to account for the specific nature of the function being analyzed?

-   Modify `-fixfn` to take two addtional optional arguments
    -   an equality test function, which defaults to `equal`, but which could
        be a fuzzy comparison of floats.
    -   an iteration limit function, which defaults to a simple counter
        with a default max value. The caller could alter the max value or
        specify some more sophisticated convergence test.
-   Modify the tests for `-fixfn` so that they use a fuzzy comparsion
    for floating point numbers such as the one given in the Emacs manual
    (with a narrower tolerance)
    
    ```emacs-lisp
    (defvar fuzz-factor 1.0e-15)
    (defun approx-equal (x y)
      (or (= x y)
          (< (/ (abs (- x y))
    	    (max (abs x) (abs y)))
    	 fuzz-factor)))
    ```

If you agree on this approach, I can work on this futher and submit a pull request.

**Comment by Fuco1** [comment]
I think the optional comparision function is a good idea.

I think magnars also run into this issue (his primary machine is mac, while mine is linux).  If we can't get the test to work stabely, it should probably be replaced with something else.

**Comment by magnars** [comment]
:+1:

**Comment by occidens** [comment]
Ok, to follow up on this: I just submitted a PR for a first pass at solving the underlying problem. Let me know what you think.

Because there are a lot of subtleties to float comparison, and dash isn't focused on numerical analysis, I tried to leave as much as possible to the caller to determine how numbers should be compared and if and when to terminate the fixpoint iteration. For this reason, I only added the basic approximate comparison test to `examples-to-tests.el`, so that the tests can work. In this way the package itself remains agnostic on how floats should be compared.

By the way, the problem with the `cos' function on OSX is strange: its fixpoint iteration doesn't even converge to within machine epsilon, hence the slightly relaxed value of 1e-15 that I used in the tests. The other example, which only has addition and exponentiation, *does* converge to a value that can be compared with equal.
